### PR TITLE
Heartbeat Scheduler Tests

### PR DESCRIPTION
**Pull request by andrewvc**
This commit aims to improve test coverage of these important packages.

It also makes some tiny changes to scheduler.go to improve understanding of some of the code.

### CODE REVIEW

**Comment by andrewvc** [comment]
This is now ready for review. @urso , if you have the time I wouldn't mind your feedback.

**Comment by andrewvc** [comment]
Pushed some fixes for the race condition

File: heartbeat/scheduler/schedule/schedule_test.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+			nil,
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := Parse(tt.schedStr)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Parse() error = %v, wantErr %v", err, tt.wantErr)
+				return
```
Comment (urso) [inline heartbeat/scheduler/schedule/schedule_test.go]: you can use `t.Fatalf` instead of `Errorf+return`.

File: heartbeat/scheduler/scheduler.go [hunk 2]
```diff
@@ -21,22 +21,29 @@ import (
...
+	statePreRunning = int32(iota + 1)
+	stateRunning
+	stateDone
+)
+
 type Scheduler struct {
-	limit   uint
-	running bool
+	limit uint
+	state *int32
```
Comment (urso) [inline heartbeat/scheduler/scheduler.go]: I guess that's an atomic? Why do we need this to be a pointer?

Consider `running atomic.Bool`. The libbeat package provides some more typed atomics.

File: heartbeat/scheduler/scheduler.go [hunk 3]
```diff
@@ -21,22 +21,29 @@ import (
...
 
 	location *time.Location
 
 	jobs   []*job
 	active uint // number of active entries
 
-	add, rm  chan *job
-	finished chan taskOverSignal
+	addCh, rm chan *job
+	finished  chan taskOverSignal
```
Comment (urso) [inline heartbeat/scheduler/scheduler.go]: for naming consistency let's also add `Ch` to `rm` and `finished` ?

File: heartbeat/scheduler/scheduler.go (hunk 2 above)
Comment (urso) [reply heartbeat/scheduler/scheduler.go]: I see you introduced 3 states. How about `atomic.Int`?

**Comment by urso** [comment]
LGTM, only some cosmetical requests. Having `state *int32` is not really required.

File: heartbeat/scheduler/scheduler.go (hunk 2 above)
Comment (andrewvc) [reply heartbeat/scheduler/scheduler.go]: ++ on the atomic Int we have.

I used a pointer because the atomic operations in the stdlib require a pointer.

**Comment by andrewvc** [comment]
Hey, thanks for the feedback. I think it all makes sense.

One thought. The concurrency situation could be simplified further if we simply removed `Start()` and had the initializer start the run goroutine. That way, adding jobs would always be through the channel. That would remove one state and we'd only need a boolean again.

@urso WDYT? Should I do that refactor?

File: heartbeat/scheduler/scheduler.go (hunk 3 above)
Comment (andrewvc) [reply heartbeat/scheduler/scheduler.go]: ++

File: heartbeat/beater/manager.go [hunk 4]
```diff
@@ -29,6 +29,8 @@ import (
 	"github.com/elastic/beats/libbeat/logp"
 	"github.com/elastic/beats/libbeat/processors"
 
+	pkgerrors "github.com/pkg/errors"
```
Comment (ruflin) [inline heartbeat/beater/manager.go]: Imports fun :-)

@urso I assume you separate libbeat and heartbeat in the past in the imports. Is this something you think ew should do? I think in most place we have all the beats imports together.

File: heartbeat/beater/manager.go [hunk 5]
```diff
@@ -231,7 +233,13 @@ func (m *monitor) Update(configs []*common.Config) error {
 		}
 
 		job := t.createJob(client)
-		jobCancel := m.manager.jobControl.Add(t.config.Schedule, id, job)
+		jobCancel, err := m.manager.jobControl.Add(t.config.Schedule, id, job)
+		if err == scheduler.ErrAlreadyStopped {
+			logp.Info("Will not add new task to scheduler. Heartbeat is stopping.")
+		} else if err != nil {
+			panic(pkgerrors.Wrap(err, "an unexpected error occurred adding a task to the scheduler: "))
```
Comment (ruflin) [inline heartbeat/beater/manager.go]: do we really want to have a panic here, meaning stopping heartbeat. Based on the current code path I think it's not possible but what if someone modifies jobControl and returns an additional error type?

File: heartbeat/scheduler/scheduler.go [hunk 6]
```diff
@@ -105,27 +113,32 @@ func NewWithLocation(limit uint, location *time.Location) *Scheduler {
 }
 
 func (s *Scheduler) Start() error {
-	if s.running {
-		return errors.New("scheduler already running")
+	if !s.transitionRunning() {
+		return errors.New("scheduler already state")
```
Comment (ruflin) [inline heartbeat/scheduler/scheduler.go]: should the current state be part of this error?

File: heartbeat/scheduler/scheduler_test.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+import (
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// The time in the island of tarawa 🏝. Good test TZ because it's pretty rare for a local box
```
Comment (ruflin) [inline heartbeat/scheduler/scheduler_test.go]: Just learned a new time zone :-)

File: heartbeat/scheduler/scheduler.go (hunk 6 above)
Comment (andrewvc) [reply heartbeat/scheduler/scheduler.go]: Yes, looks like there was a search and replace I did for `s/started/state` hence some weird grammar in a number of comments in this patch.

File: heartbeat/beater/manager.go (hunk 5 above)
Comment (andrewvc) [reply heartbeat/beater/manager.go]: Good point. Maybe I'll just make this a `logp.Error`?

**Comment by andrewvc** [comment]
@ruflin WDYT about changing the scheduler to just initialize with a running goroutine as I mentioned above? That would simplify the state machine down to two states.

**Comment by urso** [comment]
> One thought. The concurrency situation could be simplified further if we simply removed Start() and had the initializer start the run goroutine. That way, adding jobs would always be through the channel. That would remove one state and we'd only need a boolean again.

I'm no big fan of the Start/Stop pattern in general. Only advantage of this pattern is that you can store an instance in some other struct and modify/defer startup/shutdown order of sub-systems. We need this in a few places in beats, as sometimes we just initialize an object to check if the config is somewhat valid. I hope to implement better config validation in the future, the we can get rid of this pattern.

For most sub-systems the start/stop pattern is rarely required and is better be avoided if possible. Still, we must be able to cleanly shutdown sub-systems. That is, either we expose the Run method or have the constructor start the go-routine. Either way, we need some mechanism to shutdown the sub-systems. E.g. by passing some shutdown-context object to the constructor.

I'm not a big fan of `context.Context`, as I'd prefer something that is concerned about shutdown only + a well type beat context object that should be defines in the `libbeat/beat` package. There is an experimental PR trying to add some Done-context that was closed. See [done.go](https://github.com/elastic/beats/pull/2019/files#diff-6fb8aaccdd4d643111c606d8c5e9d44cR11) in #2019. I tried to make the type compatible to components implementing 'Stop' and components asking for a Done channel.

> That way, adding jobs would always be through the channel. That would remove one state and we'd only need a boolean again.

I don't fully get this comment.
1. We still want proper shutdown of the go-routine. How would this be implemented? Keep in mind that closing a buffered channel does not immediately close the channel. Sending to closed buffered channel will trigger a panic. Receiving from a closed buffered channel will still return the channel contents. The close is only obvservable by the receiver once the channel is exhausted. That's why the select+done-channel pattern is so common.
2. Please do not pass channels between subsystems.

**Comment by andrewvc** [comment]
Thanks for the thoughtful comment @urso . I'll do the refactor with the design considerations you listed in mind. WRT the 'always go through the channel' thing, I basically meant that we would keep what we have today, but the branch for 'not yet started' would just return an error. So I think we're seeing eye to eye there.

One thought though, I don't think we need to buffer the 'add' channel do we? We would only use an unbuffered channel and only write to it after checking that the state is not `Done`, which would be an atomic boolean.

**Comment by ruflin** [comment]
In general I would prefer to get small chunks of code in. The main change and purpose I see of this PR is to add additional testing to the existing code. We can do refactoring to Start / Stop in a follow up PR. Like this we also see if the tests we created for the existing code will also work for the new code.

File: heartbeat/scheduler/scheduler.go [hunk 8]
```diff
@@ -23,20 +23,27 @@ import (
 	"sync"
 	"time"
 
+	batomic "github.com/elastic/beats/libbeat/common/atomic"
```
Comment (ruflin) [inline heartbeat/scheduler/scheduler.go]: Was there a conflict just using atomic?

File: heartbeat/beater/manager.go [hunk 9]
```diff
@@ -72,7 +71,7 @@ type monitorTaskConfig struct {
 }
 
 type jobControl interface {
-	Add(sched scheduler.Schedule, name string, f scheduler.TaskFunc) func() error
+	Add(sched scheduler.Schedule, name string, f scheduler.TaskFunc) (stopFn func() error, err error)
```
Comment (urso) [inline heartbeat/beater/manager.go]: Small note, no need to address this in this PR.

It's more like cancelFn. Idea was to have some kind of handle which allows you to control the job once it is registered. I regret for having a function return here. We should introduce an interface returning a well defined Handle (e.g. `type Handle interface { Cancel(); ... }` )

File: heartbeat/scheduler/scheduler.go (hunk 8 above)
Comment (urso) [reply heartbeat/scheduler/scheduler.go]: looks like `sync/atomic` was still imported while this import was added. grepping new `scheduler.go` there should be no more naming conflict by now.

**Comment by andrewvc** [comment]
Just fixed the import issue, should be good to go

**Comment by andrewvc** [comment]
jenkins, retest this please

**Comment by andrewvc** [comment]
jenkins, retest this please

**Comment by andrewvc** [comment]
@ruflin tests keep failing, but they consistently seem to be unrelated.

**Comment by ruflin** [comment]
@andrewvc Go for it.

**Pull request merged**